src/raftstore/store/worker/read.rs

```diff
@@ -0,0 +1,884 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{self, Display, Formatter};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use kvproto::errorpb;
+use kvproto::metapb;
+use kvproto::raft_cmdpb::{CmdType, RaftCmdRequest, RaftCmdResponse};
+use mio;
+use prometheus::local::LocalHistogram;
+use rocksdb::DB;
+
+use raftstore::errors::RAFTSTORE_IS_BUSY;
+use raftstore::store::msg::Callback;
+use raftstore::store::store::Store;
+use raftstore::store::util::{self, LeaseState, RemoteLease};
+use raftstore::store::{
+    cmd_resp, Msg as StoreMsg, Peer, ReadExecutor, ReadResponse, RequestInspector, RequestPolicy,
+};
+use raftstore::Result;
+use util::collections::HashMap;
+use util::time::duration_to_sec;
+use util::timer::Timer;
+use util::transport::{NotifyError, Sender};
+use util::worker::{Runnable, RunnableWithTimer};
+
+use super::metrics::*;
+
+/// A read only delegate of `Peer`.
+#[derive(Debug)]
+pub struct ReadDelegate {
+    region: metapb::Region,
+    peer_id: u64,
+    term: u64,
+    applied_index_term: u64,
+    leader_lease: Option<RemoteLease>,
+
+    tag: String,
+}
+
+impl ReadDelegate {
+    fn from_peer(peer: &Peer) -> ReadDelegate {
+        let region = peer.region().clone();
+        let region_id = region.get_id();
+        let peer_id = peer.peer.get_id();
+        ReadDelegate {
+            region,
+            peer_id,
+            term: peer.term(),
+            applied_index_term: peer.get_store().applied_index_term,
+            leader_lease: None,
+            tag: format!("[region {}] {}", region_id, peer_id),
+        }
+    }
+
+    fn update(&mut self, progress: Progress) {
+        if let Some(region) = progress.region {
+            self.region = region;
+        }
+        if let Some(term) = progress.term {
+            self.term = term;
+        }
+        if let Some(applied_index_term) = progress.applied_index_term {
+            self.applied_index_term = applied_index_term;
+        }
+        if let Some(lease) = progress.leader_lease {
+            self.leader_lease = Some(lease);
+        }
+    }
+
+    fn handle_read(&self, req: &RaftCmdRequest, kv_engine: &Arc<DB>) -> Option<ReadResponse> {
+        if let Ok(mut resp) =
+            ReadExecutor::new(&self.region, kv_engine, &self.tag).execute(req, false)
+        {
+            if let Some(ref lease) = self.leader_lease {
+                if lease.inspect(None) == LeaseState::Valid {
+                    // Leader can read local if and only if it is in lease.
+                    cmd_resp::bind_term(&mut resp.response, self.term);
+                    return Some(resp);
+                }
+            }
+        }
+        None
```

Comment:
Seems we throw away the error information here. Why does this function return an option instead of a result?

Reply:
Local reader cant handle all requests, eg, leader steps down. When it returns `None`, caller needs to send to the request to raftstore.